### PR TITLE
Data_Engine: Handle nulls

### LINKED ISSUE #2466: Data_Engine - handle nulls
For the following methods:

 - [x] BH.Engine.Data.Compute.FilterData(FilterRequest request, IEnumerable<IBHoMObject> objects)
 - [x] BH.oM.Data.Collections.DiscretePoint.DiscretePoint(Point point, Double step)
 - [x] BH.Engine.Data.Query.Axes(Table table)
 - [x] BH.Engine.Data.Query.AxisExists(Table table, String axis)
 - [x] BH.Engine.Data.Query.AxisExists(Table table, List<String> axes)
 - [x] BH.Engine.Data.Query.IsInRange(DomainBox box1, DomainBox box2, Double tolerance)
 - [x] BH.Engine.Data.Query.IsInRange(Domain domain1, Domain domain2, Double tolerance)
 - [x] BH.Engine.Data.Query.IsInRange(Domain domain, Double val, Double tolerance)
 - [x] BH.Engine.Data.Query.FurthestSquareDistance(DomainBox box1, DomainBox box2)
 - [x] BH.Engine.Data.Query.FurthestTightSquareDistance(DomainBox box1, DomainBox box2)
 - [x] BH.Engine.Data.Query.FirstValueAbove(Table table, List<String> axes, List<Object> values, String sortAxis, Boolean allowEqual)
 - [x] BH.Engine.Data.Query.FirstValueAt(Table table, String expression, String sortOrder)
 - [x] BH.Engine.Data.Query.Filter(IEnumerable<IBHoMObject> objects, FilterRequest request)
 - [x] BH.Engine.Data.Query.Requests(LogicalAndRequest request)
 - [x] BH.Engine.Data.Query.Requests(LogicalOrRequest request)
 - [x] BH.Engine.Data.Query.Requests(LogicalNotRequest request)
 - [x] BH.Engine.Data.Query.SquareDistance(DomainBox box1, DomainBox box2)
 - [x] BH.Engine.Data.Query.ValuesAbove(Table table, List<String> axes, List<Object> values, Boolean allowEqual)
 - [x] BH.Engine.Data.Query.ValuesAbove(Table table, List<String> axes, List<Object> values, String sortAxis, Boolean allowEqual)
 - [x] BH.Engine.Data.Query.ValuesAt(Table table, List<String> axes, List<Object> values)
 - [x] BH.Engine.Data.Query.ValuesAt(Table table, String expression)
 - [x] BH.Engine.Data.Query.ValuesAt(Table table, String expression, String sortOrder)
 - [x] BH.Engine.Data.Query.Values(Table table)

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### Issues addressed by this PR
Fixes #2466

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2362663582

**Comment by FraserGreenroyd** [comment]
@BHoMbot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMbot check dataset-compliance

File: Data_Engine/Query/FirstValueAbove.cs [hunk 1]
```diff
@@ -45,6 +45,12 @@ public static partial class Query
         [Output("Data", "The data matching the provided axes and values as CustomObjects.")]
         public static CustomObject FirstValueAbove(this Table table, List<string> axes, List<object> values, string sortAxis, bool allowEqual = true)
         {
+            if(table == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot get the first value above from a null table.");
+                return new CustomObject();
```
Comment (pawelbaran) [inline Data_Engine/Query/FirstValueAbove.cs:51]: Why so?
```suggestion
                return null;
```

File: Data_Engine/Query/FirstValueAt.cs [hunk 2]
```diff
@@ -44,6 +44,12 @@ public static partial class Query
         [Output("Data", "The data matching the provided axes and values as CustomObjects.")]
         public static CustomObject FirstValueAt(this Table table, string expression, string sortOrder)
         {
+            if (table == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Cannot get the first value at from a null table.");
+                return new CustomObject();
```
Comment (pawelbaran) [inline Data_Engine/Query/FirstValueAt.cs:50]: Why so?
```suggestion
                return null;
```

**Review by pawelbaran — CHANGES_REQUESTED** [review]
In general similar thoughts as in #2465, on top of that I am concerned about returning a `CustomObject` instead of `null` - please see the code comments.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Review by pawelbaran — APPROVED** [review]
I have no objections outside of what I have mentioned in my previous comments in previous PRs of this type.

As mentioned in #2465, I am a bit concerned about the creep that these changes may possibly introduce, but it seems that the only way to check it is to merge and see how many weird behaviours bubble up.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check dataset-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2410470207

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a CI/CD instruction. I am authorising dispensation to be granted on check ref. 2410471166

**Pull request merged**